app/src/lib/api.ts

```diff
@@ -1058,7 +1058,13 @@ export class API {
       options.reloadCache
     )
 
-    if (response.status === 401) {
+    // Only consider invalid token when the status is 401 and the response has
+    // the X-GitHub-Request-Id header, meaning it comes from GH(E) and not from
+    // any kind of proxy/gateway. For more info see #12943
+    if (
+      response.status === 401 &&
+      response.headers.has('X-GitHub-Request-Id')
```

Comment:
This is slightly outside of the scope of this PR but I wonder if we should exclude the OTP response here as that's the only other response which will return a 401 while not for that matter meaning that the token should be cleared. See

https://github.com/desktop/desktop/blob/3da974a2ee54c76d5a8931caf9b42f90ede155fc/app/src/lib/api.ts#L1226-L1227

```suggestion
      response.headers.has('X-GitHub-Request-Id') &&
      !response.headers.has('X-GitHub-OTP')
```

Reply:
Sounds great, thanks!! I don't think it's completely mandatory, since when this happens there is no account to sign in again, and this check will prevent the prompt entirely: https://github.com/desktop/desktop/blob/3da974a2ee54c76d5a8931caf9b42f90ede155fc/app/src/lib/stores/app-store.ts#L539-L543

However, it'd be semantically more correct, so I think we should include it!